<a href="https://colab.research.google.com/github/aadityasomani/Aadi/blob/master/Lesson_63_Mathematical_Properties_of_Sequences_Aditya.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lesson 63: Mathematical Properties of Sequences


---

### Teacher-Student Tasks

In this class, we will learn how to calculate the general term of a sequence and compute the sum of n-terms of a sequence.

---

#### What is a Sequence?

A **sequence** is something that occurs one after another.

For example:
- The rising of the sun and its setting is a sequence. 
- Planets revolve around the sun in a sequence.

There are many such examples of sequences in our day-to-day life. However, we will learn sequences only  with numbers.

A sequence of numbers is an **ordered list** or **set of numbers**. 

For example: `2, 4, 6, 8, 10, 12, 14, 16, 18,...`

There are different types of number sequences. Let's investigate the following most widely used types of number sequences:

1. Arithmetic Sequence.
2. Geometric Sequence.
3. Fibonacci Sequence.



---

#### Task 1: Finding the $n^{th}$ term of an Arithmetic Sequence

An **arithmetic sequence** is a sequence in which the difference of consecutive terms is constant (also known as a **common difference**). This common difference is often denoted by $d$. Such arithmetic sequence is also known as  **Arithmetic progression**.

For Example:

1. Set of even numbers: `2, 4, 6, 8, 10, 12, 14, 16, 18,...`  
2. Set of natural numbers: `1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,...` 

To generate the terms of an arithmetic sequence, we keep adding the common difference $d$, to the last known term. That is if $a_1$ represents the first term and $d$ is a common difference:

$$a_2 = a_1 + d$$
$$a_3 = a_2 + d = a_1 + 2d$$
$$a_4 = a_3 + d = a_1 + 3d$$

We write the general (or $n^\text{th}$) term of a sequence as:

$$a_n = a_1 + (n - 1 ) d$$

This is the formula to be used to find the general (or $n^\text{th}$) term of an arithmetic sequence.

**Example:** *Find the general (or $n^{th}$) term of the arithmetic sequence:
2, 5, 8, 11, 14, 17, 20, 23...*


We know that the difference between any two consecutive terms is a constant, i. e. $d$. Therefore, $d$ = 5 - 2 = 3.

Placing $d$ = 3 along with the first term, $a_1$ = 2 into the general formula, we get the following:

$$a_n = 2 + (n - 1) ( 3)$$
$$a_n = 2 + 3n - 3 = 3n - 1$$

**General (or $n^{th}$) term:** $a_n= 3n - 1$

Suppose you want to find the $2^{nd}$ term in this series. Simply replace $n$ by 2,

$$a_2= 3 \times 2 - 1 = 5$$

As you may notice the $2^{nd}$ term in the above series is indeed $5$.


Let's write Python code that determines the $n^{th}$ term of an arithmetic sequence:






In [2]:
# S1.1: Code to find the nth term of an arithmetic sequence.
a1=int(input("Enter First Term"))
a2=int(input("Enter Second Term"))
n=int(input("Enter nth term"))

# Apply formula: 'an = a1 + (n−1)d' to find nth term
d=a2-a1
print(f'{n}th Term is :{a1+(n-1)*d}')

Enter First Term5
Enter Second Term7
Enter nth term3
3th Term is :9


Thus, we calculated the $n^{th}$ term of an arithmetic sequence by applying the formula in Python.

---

#### Task 2: Finding $n^{th}$ Term of a Geometric Sequence


A **geometric sequence**, also known as a **geometric progression**, is an ordered list of numbers in which each term after the first is found by multiplying the previous one by a constant called $r$, the common ratio of the two consecutive numbers.

For example, the sequence 2, 6, 18, 54,...  is a geometric progression with a common ratio 3. Similarly, 10, 5, 2.5, 1.25, ... is a geometric sequence with a common ratio $1\over2$.

Given a geometric sequence with the first term $a_1$ and the common ratio $r$, the $n^{th}$ (or general) term is given by
$$a_n=a_1 \times r^{n−1}$$ 


**Example:** *Find the general (or $n^{th}$) term of the geometric sequence: 3, 12, 48, ...*


We know that the common ratio $r$ is the ratio between any two adjacent numbers. Therefore, 
$$r = {12 \over 3} = 4$$

Placing $r = 4$ along with the first term, $a_1$ = 3 into the general formula, we get the following:

$$a_n=3 \times 4^{n−1}$$


**General (or $n^{th}$) term:** $a_n=3 \times 4^{n−1}$

Suppose you want to find the $2^{nd}$ term in the above series. Simply replace $n$ by 2,

The $a_2= 3 \times 4^{2−1} = 3 \times 4^{1} = 12$

You will notice the $2^{nd}$ term in the above series is indeed $12$.


Let's write Python code that determines the $n^{th}$ term of a geometric sequence:


In [4]:
# S2.1: Code to find the nth term of a geometric sequence.
a1=int(input("Enter First Term"))
a2=int(input("Enter Second Term"))
n=int(input("Enter nth term"))
r=a2/a1
# Apply formula: 'an = a1 x r^(n−1)' to find nth term
print(f'{n}th Term is :{a1+r**(n-1)}')

Enter First Term2
Enter Second Term4
Enter nth term8
8th Term is :130.0


Thus, we calculated the $n^{th}$ term of a geometric sequence by applying the formula in Python.

---

#### Task 3: Calculating the Sum of $n$-terms of Arithmetic Sequence

The sum of the terms of a sequence is called a **series**. To find the sum of the first $n$ terms of an arithmetic sequence, we use the following formula:



$$S_n = \frac{n \times(2a_1 + (n-1)d)}{2}$$

Where:
- $a_1$: First term of the series.
- $d$: Common difference between two consecutive numbers.
- $n$: Number of terms in the series.

**Example:**

```
Find the sum of the first 10 terms of the arithmetic series if the first term is 4 and a common difference is 3.
```

In this case:
- $n=10$
- $a_1=4$ 
- $d=3$

$$S_{10} = \frac{10\times(2 \times 4 + (10-1)\times 3)}{2}$$

$$=\frac{10\times(8 + 9 \times 3)}{2}=\frac{10\times(8 + 27)}{2}= \frac{350}{2}=175$$ 

**Sum of first 10 terms of the series** = $175$

Let's write Python code to calculate the sum of the Arithmetic Progression series:



In [9]:
# S3.1: Python Program to find Sum of Arithmetic Progression Series.
a1=int(input("Enter First Term"))
a2=int(input("Enter Second Term"))
n=int(input("Enter nth term"))

d=a2-a1
print(f'Sum Of {n}th Term is :{(n*(2*a1+((n-1)*d)))/2}')


Enter First Term2
Enter Second Term4
Enter nth term3
Sum Of 3th Term is :12.0


Thus, in this way, we can determine the sum of an Arithmetic Progression series by applying the formula in Python.

---

#### Task 4: Calculating the Sum of $n$-terms of Geometric Sequence

To find the sum of the first $n$ terms of a geometric sequence, we use the following formula:

$$S_n = \frac{a_1 \times (1-r^n)}{(1- r)}$$

Where:
- $a_1$: First term of the series.
- $r$: Common ratio between two consecutive numbers. 
- $n$: Number of terms in the series.


**Example:**

```
Find the sum of the first 3 terms of the geometric series if the first term is 1 and a common ratio is 4.
```

In this case:
- $n=3$
- $a_1=10$ 
- $r=3$

$$S_{3} = \frac{10 \times (1-3^{3})}{(1- 3)}$$
$$=\frac{10 \times  (1-27)}{-2}=\frac{10 \times (-26)}{-2}=130$$

**Sum of first 3 terms of the series** = $130$


Let's write Python code that calculates the sum of the Geometric Progression series:

In [10]:
# S4.1: Python Program to find the sum of geometric progression series
a1=int(input("Enter First Term"))
a2=int(input("Enter Second Term"))
n=int(input("Enter nth term"))
r=a2/a1
# Apply formula: 'an = a1 x r^(n−1)' to find nth term
print(f'Sum of {n} Term is :{(a1*(1-(r**n)))/(1-r)}')
# Apply formula sum = (a * (1 - r^ n)) / (1- r)



Enter First Term3
Enter Second Term6
Enter nth term6
Sum of 6 Term is :189.0


Thus, in this way, we can determine the sum of a Geometric Progression series by applying the formula in Python.


---

#### Task 5: Calculating the $n^{th}$ Term and the Sum of $n$-terms of a Fibonacci Sequence

Fibonacci sequence is a very interesting ordered arrangement of numbers where the terms are found by adding the two previous terms. 

It looks like this:  **0, 1, 1, 2, 3, 5, 8, 13, 21, 34, ...**

  The first two terms are 0 and 1. All other terms are obtained by adding two preceding terms. This means to say the $n^{th}$ term is the sum of $(n-1)^{th}$ and $(n-2)^{th}$ term.

**For example:** 

- The third term `1` is obtained by adding the first term `0` and the second term `1`.

- Similarly, the fifth term `3` is obtained by adding the third term `1` and the fourth term `2`.

- The sum of first 5 terms would be: `0 + 1 + 1 + 2 + 3 = 7`

Let us calculate the $n^{th}$ term in a Fibonacci sequence and also calculate the sum of the first $n$-terms using the steps given below in Python:

1. Create the following two variables which signify the first two numbers of the sequence:
    - `first_num = 0`
    - `second_num = 1`

2. Create a variable `fibonacci_sum` to store the sum of the first $n$-terms and initialize it with `1`.

3. Ask the user to enter the term they would like to find and store it in a variable `n`. Convert it into integer using `int()` function.

4. Create a condition such that, if `n==1`, then print `"The number in position 1 is: 0"`, as the Fibonacci sequence always starts with `0`.

5. Else start a `for` loop which ranges from `2` to `n`. Inside this `for` loop:
 - Create a variable `temp` which  holds the sum of the preceding two terms.
 - Assign the value of `second_num` to `first_num` and `temp` to `second_num` to change the values of the preceding two terms for the next iteration.
 - Add the value of `second_num` to `fibonacci_sum` in each iteration as follows: 
   `fibonacci_sum = fibonacci_sum + second_num`.

6. Outside for loop, print the $n^{th}$ term in the following way:

  `"The number in position", n ,"is:", second_num`, where `second_num` holds the $n^{th}$ term.

7. Print the sum of the first $n$-terms of the Fibonacci series by printing the value of the variable `fibonacci_sum`.

In [21]:
# S5.1:Python Program to find the nth term and sum of fibonacci series
first_num = 0
second_num = 1
sum=1
n=int(input("Enter nth terms:"))
if (n==1):
    print("The number in position 1 is: 0")
else:
    for i in range(2,n):
        temp=first_num+second_num
        first_num=second_num
        second_num=temp
        sum+=second_num
    print("The number in position", n ,"is:", second_num)
    print("Sum of Fibonacci is :",sum)
        

Enter nth terms:4
The number in position 4 is: 2
Sum of Fibonacci is : 4


In [29]:
def n_fibonacci(n,a1,a2):
    if n-1==0:
        return a1
    else:
        temp=a1+a2
        a1=a2
        return fibonacci(n-1,a1,temp)
n_fibonacci(9,0,1)

21

In [35]:
first_num = 0
second_num = 1
sum=1
n=int(input("Enter nth terms:"))
if (n==1):
    print("0")
else:
    print("0")
    for i in range(2,n+1):
        temp=first_num+second_num
        first_num=second_num
        second_num=temp
        sum+=second_num
        print(first_num)

Enter nth terms:8
0
1
1
2
3
5
8
13


Thus, we obtained the $n^{th}$ term of a Fibonacci series and also calculated the sum of the first $n$-terms.


Hence, in this class, we learned the types of sequence and also how to calculate the general term of a sequence and compute the sum of $n$-terms of a sequence using Python. In the next class, we will explore a linear data structure called Linked List.